File: Sources/structlyC/04_linked_lists/linked_list_values.cpp
//
//  linked_list_values.cpp
//  MyCLI
//
//  Created by new on 3/15/26.
//

#include <string>
#include <vector>

class Node {
public:
    std::string val;
    Node *next;

    Node(std::string initialVal) {
        val = initialVal;
        next = nullptr;
    }
};

std::vector<std::string> linkedListValuesIter(Node *head) {
    Node *node = head;
    std::vector<std::string> vals{};
    while (node != nullptr) {
        vals.push_back(node->val);
        node = node->next;
    }
    return vals;
}

void linkedListValuesRec(Node *head, std::vector<std::string> &values) {
    if (head == nullptr) {
        return;
    }
    values.push_back(head->val);
    linkedListValuesRec(head->next, values);
}

std::vector<std::string> linkedListValuesRec(Node *head) {
    std::vector<std::string> values{};
    linkedListValuesRec(head, values);
    return values;
}

File: Sources/structly/04_linked_lists/LinkedListValues.swift
//
//  LinkedListValues.swift
//  MyCLI
//
//  Created by new on 3/15/26.
//

func linkedListValues2(_ head: Node<String>?) -> [String] {
    var vals = [String]()
    var node = head
    while node != nil {
        vals.append(node!.val)
        node = node?.next
    }
    return vals
}

func linkedListValues(_ head: Node<String>?) -> [String] {
    func rec(node: Node<String>?, acc: inout [String]) {
        guard let node else {
            return
        }
        acc.append(node.val)
        rec(node: node.next, acc: &acc)
    }
    var acc: [String] = []
    rec(node: head, acc: &acc)
    return acc
}
